File: java/src/test/java/othello/ModelsTest.java
package othello;

import java.util.List;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import othello.Models.Direction;
import othello.Models.Disk;
import othello.Models.Move;
import othello.Models.Square;
import othello.Models.Step;

import static org.junit.jupiter.api.Assertions.*;

public class ModelsTest {

    @Test
    public void stepAddition() {
        Step result = new Step(0, 0).add(new Step(1, 1));
        assertEquals(new Step(1, 1), result);

        result = new Step(-1, 0).add(new Step(1, 0));
        assertEquals(new Step(0, 0), result);

        result = result.add(new Step(-1, -1));
        assertEquals(new Step(-1, -1), result);

        result = result.add(new Step(1, 1));
        assertEquals(new Step(0, 0), result);
    }

    @Test
    public void squareAddition() {
        Square result = new Square(4, 4).add(new Square(1, 1));
        assertEquals(new Square(5, 5), result);

        result = new Square(4, 4).add(new Square(0, 0));
        assertEquals(new Square(4, 4), result);

        result = new Square(4, 4).add(new Step(-1, 1));
        assertEquals(new Square(3, 5), result);

        result = result.add(new Square(0, 0));
        assertEquals(new Square(3, 5), result);

        Square another = new Square(-3, -2).add(new Square(2, 3));
        assertEquals(new Square(-1, 1), another);

        result = result.add(new Step(-1, -1));
        assertEquals(new Square(2, 4), result);

        result = result.add(new Step(-1, -1));
        assertEquals(new Square(1, 3), result);

        result = result.add(new Step(-1, -1));
        assertEquals(new Square(0, 2), result);

        result = result.add(new Step(-1, -1));
        assertEquals(new Square(-1, 1), result);

        result = result.add(new Step(1, -1));
        assertEquals(new Square(0, 0), result);

        result = result.add(new Step(-1, -1));
        assertEquals(new Square(-1, -1), result);
    }

    @ParameterizedTest
    @CsvSource({
        "-1, -1, 0, 0",
        "-1,  0, 0, 1",
        "-1,  1, 0, 2",
        " 0, -1, 1, 0",
        " 0,  1, 1, 2",
        " 1, -1, 2, 0",
        " 1,  0, 2, 1",
        " 1,  1, 2, 2"
    })
    public void stepDirections(int stepX, int stepY, int expectedX, int expectedY) {
        Square origin = new Square(1, 1);
        Step step = new Step(stepX, stepY);
        Square expected = new Square(expectedX, expectedY);
        assertEquals(expected, origin.add(step));
    }

    @Test
    public void diskBoardChar() {
        assertEquals("B", Disk.BLACK.boardChar());
        assertEquals("_", Disk.EMPTY.boardChar());
        assertEquals("W", Disk.WHITE.boardChar());
    }

    @Test
    public void diskOpponent() {
        assertEquals(Disk.WHITE, Disk.BLACK.opponent());
        assertEquals(Disk.EMPTY, Disk.EMPTY.opponent());
        assertEquals(Disk.BLACK, Disk.WHITE.opponent());
    }

    @Test
    public void moveLogEntry() {
        Move b = new Move(
            new Square(3, 2),
            Disk.BLACK,
            10,
            List.of(new Direction(new Step(1, 0), 10))
        );
        assertEquals("B:(3,2),10", b.logEntry());

        Move w = new Move(
            new Square(0, 0),
            Disk.WHITE,
            1,
            List.of(new Direction(new Step(1, 0), 1))
        );
        assertEquals("W:(0,0),1", w.logEntry());
    }
}

File: kotlin-native/src/nativeTest/kotlin/othello/ModelsTest.kt
package othello

import kotlin.test.Test
import kotlin.test.assertEquals

class ModelsTest {
    @Test
    fun stepAddition() {
        var result = Step(0, 0) + Step(1, 1)
        assertEquals(Step(1, 1), result)

        result = Step(-1, 0) + Step(1, 0)
        assertEquals(Step(0, 0), result)

        result += Step(-1, -1)
        assertEquals(Step(-1, -1), result)

        result += Step(1, 1)
        assertEquals(Step(0, 0), result)
    }

    @Test
    fun squareAddition() {
        var result = Square(4, 4) + Square(1, 1)
        assertEquals(Square(5, 5), result)

        result = Square(4, 4) + Square(0, 0)
        assertEquals(Square(4, 4), result)

        result = Square(4, 4) + Step(-1, 1)
        assertEquals(Square(3, 5), result)

        result += Square(0, 0)
        assertEquals(Square(3, 5), result)

        val another = Square(-3, -2) + Square(2, 3)
        assertEquals(Square(-1, 1), another)

        result += Step(-1, -1)
        assertEquals(Square(2, 4), result)

        result += Step(-1, -1)
        assertEquals(Square(1, 3), result)

        result += Step(-1, -1)
        assertEquals(Square(0, 2), result)

        result += Step(-1, -1)
        assertEquals(Square(-1, 1), result)

        result += Step(1, -1)
        assertEquals(Square(0, 0), result)

        result += Step(-1, -1)
        assertEquals(Square(-1, -1), result)
    }

    @Test
    fun stepDirections() {
        val origin = Square(1, 1)
        val stepDirections = arrayOf(
            Step(-1, -1),
            Step(-1, 0),
            Step(-1, 1),
            Step(0, -1),
            Step(0, 1),
            Step(1, -1),
            Step(1, 0),
            Step(1, 1),
        )

        val expectedResults = arrayOf(
            Square(0, 0),
            Square(0, 1),
            Square(0, 2),
            Square(1, 0),
            Square(1, 2),
            Square(2, 0),
            Square(2, 1),
            Square(2, 2),
        )

        stepDirections.forEachIndexed { i, step ->
            val result = origin + step
            assertEquals(expectedResults[i], result, "Step direction: $step")
        }
    }

    @Test
    fun diskBoardChar() {
        assertEquals("B", Disk.Black.boardChar(color = false))
        assertEquals("_", Disk.Empty.boardChar(color = false))
        assertEquals("W", Disk.White.boardChar(color = false))
    }

    @Test
    fun diskOpponent() {
        assertEquals(Disk.White, Disk.Black.opponent())
        assertEquals(Disk.Empty, Disk.Empty.opponent())
        assertEquals(Disk.Black, Disk.White.opponent())
    }

    @Test
    fun moveLogEntry() {
        val b = Move(
            square = Square(3, 2),
            disk = Disk.Black,
            value = 10,
            directions = listOf(
                Direction(
                    step = Step(1, 0),
                    count = 10,
                ),
            ),
        )
        assertEquals("B:(3,2),10", b.logEntry())

        val w = Move(
            square = Square(0, 0),
            disk = Disk.White,
            value = 1,
            directions = listOf(
                Direction(
                    step = Step(1, 0),
                    count = 1,
                ),
            ),
        )
        assertEquals("W:(0,0),1", w.logEntry())
    }
}
